src/System.Memory/src/System/Buffers/Text/Base64Encoder.cs

```diff
@@ -0,0 +1,258 @@
+
+            int srcLength = bytes.Length;
+            int destLength = utf8.Length;
+
+            int sourceIndex = 0;
+            int destIndex = 0;
+            int result = 0;
+
+            if (destLength >= GetMaxEncodedToUtf8Length(srcLength))
+            {
+                while (sourceIndex < srcLength - 2)
+                {
+                    result = Encode(ref Unsafe.Add(ref srcBytes, sourceIndex));
+                    Unsafe.WriteUnaligned(ref Unsafe.Add(ref destBytes, destIndex), result);
+                    destIndex += 4;
+                    sourceIndex += 3;
+                }
+            }
+            else
+            {
+                while (sourceIndex < srcLength - 2)
+                {
+                    result = Encode(ref Unsafe.Add(ref srcBytes, sourceIndex));
+                    if (destIndex > destLength - 4) goto DestinationSmallExit;
+                    Unsafe.WriteUnaligned(ref Unsafe.Add(ref destBytes, destIndex), result);
+                    destIndex += 4;
+                    sourceIndex += 3;
+                }
+            }
+            
+            if (isFinalBlock != true) goto NeedMoreDataExit;
+
+            if (sourceIndex == srcLength - 1)
+            {
+                result = EncodeAndPadTwo(ref Unsafe.Add(ref srcBytes, sourceIndex));
+                if (destIndex > destLength - 4) goto DestinationSmallExit;
+                Unsafe.WriteUnaligned(ref Unsafe.Add(ref destBytes, destIndex), result);
+                destIndex += 4;
+                sourceIndex += 1;
+            }
+            else if (sourceIndex == srcLength - 2)
+            {
+                result = EncodeAndPadOne(ref Unsafe.Add(ref srcBytes, sourceIndex));
+                if (destIndex > destLength - 4) goto DestinationSmallExit;
+                Unsafe.WriteUnaligned(ref Unsafe.Add(ref destBytes, destIndex), result);
+                destIndex += 4;
+                sourceIndex += 2;
+            }
+
+            consumed = sourceIndex;
+            written = destIndex;
+            return OperationStatus.Done;
+
+            NeedMoreDataExit:
+            consumed = sourceIndex;
+            written = destIndex;
+            return OperationStatus.NeedMoreData;
+
+            DestinationSmallExit:
+            consumed = sourceIndex;
+            written = destIndex;
+            return OperationStatus.DestinationTooSmall;
+        }
+
+        /// <summary>
+        /// Returns the maximum length (in bytes) of the result if you were to encode binary data within a byte span of size "length".
+        /// </summary>
+        /// <exception cref="System.OverflowException">
+        /// Thrown when the specified <paramref name="length"/> is larger than 1610612733 (since encode inflates the data by 4/3).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int GetMaxEncodedToUtf8Length(int length)
+        {
+            Debug.Assert(length >= 0);
```

Comment:
Same, i.e. I think we should throw for negative lengths.

Reply:
Agreed.  Public members need to either accept their inputs (no assert, no check) or properly validate (check and throw).  Asserting on inputs is only valid for non-public members, to guard that the public member already did the validation.

Reply:
Do we care what exception type gets thrown? Personally I don't care if it's ArgumentException or OverflowException, or even if it's potentially both (negative vs. too large). It's not like the developer can catch and handle it anyway.

Reply:
ArgumentOutOfRangeException seems to match the problem space to me.  (Well, really, argument out of domain (since it's an input); but that's just my math degree talking)